Source/Dafny/AST/DafnyAst.cs

```diff
@@ -6927,8 +6927,35 @@ public Statement(IToken tok, IToken endTok)
 
     /// <summary>
     /// Returns the non-null expressions of this statement proper (that is, do not include the expressions of substatements).
+    /// Filters all sub expressions that are not part of specifications
     /// </summary>
-    public virtual IEnumerable<Expression> SubExpressions {
+    public IEnumerable<Expression> SubExpressions {
+      get {
+        foreach (var e in SpecificationSubExpressions) {
+          yield return e;
+        }
+
+        foreach (var e in NonSpecificationSubExpressions) {
+          yield return e;
+        }
+      }
+    }
+
+    /// <summary>
+    /// Returns the non-null expressions of this statement proper (that is, do not include the expressions of substatements).
```

Comment:
What is a null Expression? I think we don't need to specify that this shouldn't returns nulls.

What are specifications in the context of a statement? Specification makes me think of requires and ensures clauses. Sorry for the noob question :-)

Reply:
I just took this comment from the previous comment of "SubExpressions", and I understand it that "proper subexpressions" (that is, referenced by no other member of this expression) could sometimes be null, meaning they were not provided. We don't return them in this case, so that's why I think this comment makes sense although it should be straightforward.

Regarding the specifications in the context of a statement, let me show you the 7 cases in which it matters, for which I override `SpecificationSubExpressions` in `DafnyAST.cs`. all the following expressions E and their variants are specifications because they are not compiled.

First the four cases that can be compiled, which were regression problems, because they were previously assuming that their specifications were compiled for this PR (there was no way to distinguish Guard from E using only SubExpressions for the while loop for example)
```
for j := 0 to 5 invariant E(j) {...}
while(Guard) invariant E(j) {...}
forall j | 0 <= j < 5 ensures E {...}
modify E {...}
```

Second, the next three cases were modified accordingly to be consistent, but the sentence is marked as ghost anyway.
```
assert E1;
assume E2;
calc { E1; E2; ....}
```

Note that `expect E` is not in this category because in this case E would be executed at run-time (i.e. for tests)